cmd/pd-server/main.go

```diff
@@ -224,7 +225,36 @@ func start(cmd *cobra.Command, args []string, services ...string) {
 
 	var sig os.Signal
 	go func() {
-		sig = <-sc
+	SignalHandler:
+		for {
+			select {
+			case sig = <-sc:
+				log.Info("Received a signal", zap.String("signal", sig.String()))
+				if svr.GetMember().IsLeader() {
+					log.Info("Transferring leadership due to signal", zap.String("signal", sig.String()))
+					err := svr.GetMember().ResignEtcdLeader(context.TODO(), svr.Name(), "")
+					if err != nil {
+						log.Error("Failed to transfer leadership", zap.String("signal", sig.String()))
+					} else {
+						for i := 0; svr.GetMember().IsLeader() && i < 30; i++ {
+							time.Sleep(1 * time.Second)
+						}
+						if svr.GetMember().IsLeader() {
+							log.Error("timeout to transfer leadership", zap.String("signal", sig.String()))
+						} else {
+							log.Info("Transferred leadership")
+						}
+					}
+				}
+				switch sig {
+				case syscall.SIGHUP:
+					log.Info("got SIGHUP, waiting for next signal")
+				default:
+					log.Info("Exit signal handler")
+					break SignalHandler
```

Comment:
Looks like this `break` will bypass the function call of `cancel()` , will it cause any problem?

Reply:
Oh really?  Looks like the SignalHandler block includes the line of `cancel()`.  Breaking the `SignalHandler` will skip running the `cancel()` , right?

Reply:
Can we avoid any misunderstandings caused by this, regardless of whether you are right or I am right in the end?

Reply:
I just checked, and the `cancel()` is called, I checked by adding a log between the loop and the cancel call, and it was in the output. But I agree with you that it is not the issue here.